Support for slack events

AgarFu opened the pull request:
This PR is reusing most of the classes from the SlackRTM modules, there is no documentation nor tests in this PR and the main class can be refactored in order to be able to leverage most methods from `SlackRTMBackend` class.

Please let me know if this is something that you're interested in incorporate to the project I can add the missing bits, it is now in a "works for me" status.

nzlosh commented:
@AgarFu Nice work, thanks for taking the initiative on creating this PR.

Slack seems to be in a state of disorder with regard to the Python SDK, Application Authorisation model and API access.

From what I can tell, the current situation is as follows:

## Python SDK
  - Comes in two versions v1 (used by slack backend) and v2 (used by slackRTM backend).
  - Active development on v1 has been halted in favour of v2.

## Slack API
  - RTM - Real Time Messaging (Websocket (client to server) receive all events from the workspace as well as ability to send events to Slack)
  - Web API - Post events to Slack.
  - Event API - Subscribe to specific events (server connects to client).  Push model only, must reply via Web API.

## Application Authentication
  - "Classic" App tokens that give bots "all" rights scope (strongly urged by Slack to use fine grain scope).  Has full access to the RTM and Web API.
  - New App tokens give fine grain control to what a bot can do.  These tokens can't be used with the RTM API, must use Event API and Web API.

This is situation is causing a lot of confusion for users.  What the Slack backend needs to do is have a way for the user to express the type of token they're using ("Classic" or "Fine grain") and then adapt it's behaviour accordingly.

In the case of the Event API, the errbot webserver will probably need to used for the Slack server to connection.  From what I understand, the event API events and RTM API events should be handled in the same way by errbot so we could probably mutalise code for these 2 use cases.

@sijis @AgarFu What do you think about attempting to merge RTM/Event/Web API support into a single Slack backend?

nzlosh commented:
Here is the list of events with the RTM/Event API method indicated as reference.  **https://api.slack.com/events**

AgarFu commented:
Fully agree on all that you said @nzlosh, I'm actually using errbot webserver to listen to posts from Slack. I can refactor `SlackRTMBackend` in order to be able to share even more between `SlackRTMBackend` and `SlackEventsBackend`. Right now these two backends are closer together than `SlackRTMBackend` and `SlackBackend`. That will allow a further refactor to fuse all of them into one fully capable client.

Is there a strong opposition to create a backends.slack module with all shared classes between both backends?

sijis commented:
I'm open to combining it into the SlackRTM backend. I think it logically makes sense.

AgarFu commented:
Just refactored `SlackRTMBackend` extracting `SlackBackendBase` that is doing all the heavy lifting and implementing the handlers for all events.

My new class now only overrides the `__init__` and `serve_forever` methods, I'm dynamically wrapping all possible events coming from slack to methods in the `SlackBackendBase`, this is much more clean I think.

There are no unit test for `SlackRTMBackend` (or at least I couldn't find them) so I don't know if that class is still working @sijis would be very nice if you can verify that my refactor did not break the RTM client.

AgarFu commented:
@nzlosh @sijis any more comments? do you want me to try to refactor this patch further and fuse SlacmRTM and SlackEVENTS?

sijis commented:
I will try to look into the details of this PR within the next week. I can provide feedback and more comments then.

AgarFu commented:
Hi it seems that Slack is deprecating more and more APIs, and there is interest in the community to support the slack event api. I can devote some more time if you consider that the PR needs rework.

I've rebased my branch with the latest changes in the main branch and also fixed some errors due to these deprecations.

Also the new slack_sdk python module is close to be ready, I think errbot should migrate to the new sdk sooner rather than later but that may have implications in the RTM support.

@nzlosh @sijis thoughts?

nzlosh commented:
@AgarFu I've done some work in my repo to integrate both the RTM and Events API into a single backend.  Would you be interested in pulling them into your repo to push the changes into this PR?

AgarFu commented:
Sure thing let me take a look, thx @nzlosh

AgarFu commented:
I'll integrate the changes in a different branch for now, will add some tests and then push the changes to my branch I just want to make sure the latest changes that I applied to fix deprecation errors are still in place. I didn't add the proper tests two weeks ago but this PR is substantially more aggressive and still there are no tests for the backend.

nzlosh commented:
@AgarFu I agree with you, but I'd like to see this PR merged as a MVP so we can start iterating on refactoring/unit tests/bug fixes and documentation for this new backend.

A deprecation plan has been proposed here https://github.com/errbotio/errbot/issues/1480, which I hope shows why multiple slack backends is harming errbot's user experience.

The SlackRTM backend was meant to be the successor, but with slack events and the slack client bump to v3, it lacked the features to cover all use cases.

AgarFu commented:
Ok let me confirm that my team is using a different branch and I'll merge your changes, I'd like to coordinate with you for the next iterations I'd like to at least add the testing as I said before to make sure we do not have regressions and we also have a path going forward to update to the latest versions of slack sdk

AgarFu commented:
@nzlosh your changes are merged in my branch feel free to merge them into mainstream, THX!

nzlosh commented:
@AgarFu would you mind merging the latest changes from my repository https://github.com/nzlosh/errbot/tree/slack_event_api_support into your `slack_event_api_support` branch?  I don't know why, but each time I attempt to create a PR against your repository, github doesn't display it in the list of available base repositories.

@sijis Once the changes have been merged into @AgarFu's repository, the `slackv3` backend will be ready to be merged to master.

The `slackv3` requirements include the RTM and Events python modules.  The backend can automatically detect which token type is being used and will initialise the correct client.  Not all events are supported by both API types.  However the events that errbot requires which are common to both backends end up calling the same methods.

AgarFu commented:
will do

sijis commented:
This is needing a rebase after the latest release. sorry folks.

nzlosh commented:
@AgarFu I've merged master and pushed to my branch.

sijis commented:
Looks to be applied in https://github.com/errbotio/err-backend-slackv3
